test/vgpu/test_recv.py

```diff
@@ -0,0 +1,363 @@
+        missing: bool,
+    ) -> None:
+        """Send heaps and check that they arrive and that metrics are correct.
+
+        There are two parametrisations:
+
+        missing
+            If true, do not send any heaps from batches 4:10 (covering the
+            tail of one chunk, the whole of the next, and the start of the
+            third) plus one more heap.
+        reorder
+            If true, shuffle the heaps randomly before sending them.
+        """
+        n_batches = 25
+        rng = np.random.default_rng(seed=1)
+        data = rng.integers(
+            -127, 127, size=(N_POLS, n_batches, layout.n_channels, layout.n_spectra_per_heap, COMPLEX), dtype=np.int8
+        )
+        first_chunk_id = 123
+        first_timestamp = first_chunk_id * layout.chunk_timestamp_step
+        heaps = list(gen_heaps(layout, data, first_timestamp))
+        if missing:
+            # Remove a contiguous range of heaps. This is not the most thorough
+            # test, but katgpucbf.recv.iter_chunks already gets cover from fgpu
+            # and xbgpu tests.
+            del heaps[4 * layout.batch_heaps : 10 * layout.batch_heaps + 1]
+        if reorder:
+            # Ignore is needed due to https://github.com/numpy/numpy/issues/29974
+            rng.shuffle(heaps)  # type: ignore
+
+        with PromDiff(namespace=METRIC_NAMESPACE) as prom_diff:
+            # Heap with no payload - representing any sort of metadata heap such as descriptors
+            await send_stream.async_send_heap(spead2.send.Heap(FLAVOUR))
+            for heap in heaps:
+                await send_stream.async_send_heap(heap.heap, substream_index=heap.substream_index)
+            for queue in queues:
+                queue.stop()  # Flushes out the receive stream
+
+            start_batch = 0
+            expected_timestamp = first_timestamp
+            i = 0
+            async for chunk in iter_chunks(data_ringbuffer, layout, sensors, time_converter, POL_LABELS):
+                with chunk:
+                    if missing and i == 2:
+                        # Previous chunk is entirely missing
+                        start_batch += layout.n_batches_per_chunk
+                        expected_timestamp += layout.chunk_timestamp_step
+                    n_chunk_batches = min(layout.n_batches_per_chunk, n_batches - start_batch)
+                    expected_present = np.zeros_like(chunk.present)
+                    expected_present[:, :n_chunk_batches] = 1
+                    expected_data = np.zeros_like(chunk.data)
+                    expected_data[:, :n_chunk_batches] = data[:, start_batch : start_batch + n_chunk_batches]
+                    if missing:
+                        match i:
+                            case 1:
+                                expected_present[:, 1:] = 0
+                            case 2:
+                                expected_present[:, :1] = 0
+                                expected_present[0, 1, 0] = 0
```

Comment:
I understand there is a lot of setup for data Tx and Rx here; did you have any other reasoning for consolidating the `missing` branch of this test into here?
- As far as I can tell, the other `test_recv`s have a separate `test_missing_heaps` which definitely improves readability (in my opinion)

I know we ultimately dictate how the test is done and verified, but I think this `if missing` would be more digestible if it was confined to its own test.

Reply:
I didn't want to duplicate 90% of the test and then have to maintain both copies. I think the other test_recv modules evolved a bit differently (particularly for fgpu, which changed substantially once stream-groups were introduced in spead2) which got them to where they are.